src/Security/Authentication/Negotiate/src/Internal/ReflectedNegotiateState.cs

```diff
@@ -5,38 +5,63 @@
 using System.Linq;
 using System.Net;
 using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Runtime.InteropServices;
 using System.Security.Authentication;
 using System.Security.Principal;
 
 namespace Microsoft.AspNetCore.Authentication.Negotiate
 {
     internal class ReflectedNegotiateState : INegotiateState
     {
+        // https://www.gnu.org/software/gss/reference/gss.pdf
+        private const uint GSS_S_NO_CRED = 7 << 16;
+
         private static readonly ConstructorInfo _constructor;
         private static readonly MethodInfo _getOutgoingBlob;
         private static readonly MethodInfo _isCompleted;
         private static readonly MethodInfo _protocol;
         private static readonly MethodInfo _getIdentity;
         private static readonly MethodInfo _closeContext;
+        private static readonly FieldInfo _statusCode;
+        private static readonly FieldInfo _statusException;
+        private static readonly MethodInfo _getException;
+        private static readonly FieldInfo _gssMinorStatus;
+        private static readonly Type _gssExceptionType;
 
         private readonly object _instance;
 
         static ReflectedNegotiateState()
         {
-            var ntAuthType = typeof(AuthenticationException).Assembly.GetType("System.Net.NTAuthentication");
+            var secAssembly = typeof(AuthenticationException).Assembly;
+            var ntAuthType = secAssembly.GetType("System.Net.NTAuthentication", throwOnError: true);
             _constructor = ntAuthType.GetConstructors(BindingFlags.NonPublic | BindingFlags.Instance).First();
             _getOutgoingBlob = ntAuthType.GetMethods(BindingFlags.NonPublic | BindingFlags.Instance).Where(info =>
-                info.Name.Equals("GetOutgoingBlob") && info.GetParameters().Count() == 2).Single();
+                info.Name.Equals("GetOutgoingBlob") && info.GetParameters().Count() == 3).Single();
             _isCompleted = ntAuthType.GetMethods(BindingFlags.NonPublic | BindingFlags.Instance).Where(info =>
                 info.Name.Equals("get_IsCompleted")).Single();
             _protocol = ntAuthType.GetMethods(BindingFlags.NonPublic | BindingFlags.Instance).Where(info =>
                 info.Name.Equals("get_ProtocolName")).Single();
             _closeContext = ntAuthType.GetMethods(BindingFlags.NonPublic | BindingFlags.Instance).Where(info =>
                 info.Name.Equals("CloseContext")).Single();
 
-            var negoStreamPalType = typeof(AuthenticationException).Assembly.GetType("System.Net.Security.NegotiateStreamPal");
+            var securityStatusType = secAssembly.GetType("System.Net.SecurityStatusPal", throwOnError: true);
+            _statusCode = securityStatusType.GetField("ErrorCode");
+            _statusException = securityStatusType.GetField("Exception");
+
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```

Comment:
Anything within an OS check can be removed after https://github.com/dotnet/corefx/pull/40105